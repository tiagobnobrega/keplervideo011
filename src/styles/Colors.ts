export const COLORS = {
  WHITE: '#FFFFFF',
  SMOKE_WHITE: '#F1F1F1',
  BLACK: '#000000',
  DARK_GRAY: '#1C1C1C',
  LIGHT_GRAY: '#D3D3D3',
  GRAY: '#808080',
  CHARCOAL_GRAY: '#3D3D3D',
  PALE_GRAY: '#A3A3A3',
  MEDIUM_GRAY: '#5B5B5B',
  BLUE: '#0000FF',
  RED: '#FF0000',
  YELLOW: '#FFFF00',
  LIME_GREEN: '#d9dbda',
  STRONG_RED: '#C82647',
  PALE_LIGHT_BLUE: '#F8F8FF',
  PURPLE: '#800080',
  TRANSPARENT: 'transparent',
  ORANGE: '#FFC166',
  KASHMIR_BLUE: '#4c669f',
  CHAMBRAY: '#3B5998',
  BISCAY: '#192F6A',
  PALE_BLUE_WHITE: '#EAEFFF',
};
