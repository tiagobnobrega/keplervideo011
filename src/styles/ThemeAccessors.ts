import { Theme } from '@amzn/kepler-ui-components';

/**
 * Obtains the card dimensions for md aka medium vertical card
 *
 * @param theme Theme from which card dimensions are obtained.
 * @returns card dimensions
 * @throws exception if width or height are undefined
 */
export const getVerticalCardDimensionsMd = (
  theme: Theme,
): { width: number; height: number } => {
  const width = theme.card?.container.size.width.vertical.md;
  const height = theme.card?.container.size.height.vertical.md;

  if (width === undefined || height === undefined) {
    throw new Error('Card dimensions are undefined');
  }

  return {
    width: width,
    height: height,
  };
};
