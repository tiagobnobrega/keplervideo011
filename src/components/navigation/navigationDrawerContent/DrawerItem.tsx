import { isEqual } from 'lodash';
import React from 'react';
import {
  Animated,
  GestureResponderEvent,
  StyleProp,
  StyleSheet,
  View,
  ViewStyle,
} from 'react-native';
import { COLORS } from '../../../styles/Colors';
import { scaleUxToDp } from '../../../utils/pixelUtils';

export interface DrawerItemProps {
  icon?: (props?: {
    color: string;
    size: number;
    focused: boolean;
  }) => React.ReactNode;
  label?: string;
  isSelected: boolean;
  isFocused: boolean;
  isDrawerInFocus: boolean;
  drawerWidth: number;
  routeName: string;
  onPress: (routeName: string) => void;
}

const DrawerItem = ({
  icon,
  label,
  isSelected,
  isFocused,
  isDrawerInFocus,
  drawerWidth,
  onPress,
  routeName,
}: DrawerItemProps) => {
  const Icon = icon
    ? icon({
        color: isSelected ? COLORS.WHITE : COLORS.TRANSPARENT,
        size: 2,
        focused: isFocused,
      })
    : null;

  const computedStyle = StyleSheet.create({
    drawerItemContainer: {
      width: drawerWidth - scaleUxToDp(20),
      flexDirection: 'row',
      alignItems: 'center',
      padding: scaleUxToDp(20),
      marginLeft: scaleUxToDp(10),
      marginRight: scaleUxToDp(10),
      marginBottom: scaleUxToDp(10),
    },
  });

  const viewStyles: StyleProp<ViewStyle>[] = [
    computedStyle.drawerItemContainer,
  ];
  if (isSelected && !isDrawerInFocus) {
    viewStyles.push(styles.selectedElement);
  }
  if (isDrawerInFocus && isFocused) {
    viewStyles.push(styles.focusElement);
  }

  const handleTouch = (event: GestureResponderEvent) => {
    event.preventDefault();
    onPress(routeName);
  };

  return (
    /*
      Parent takes care of focus management, therefore we would
      explicitly tell the focus manager to not give us focus.
    */
    <View
      style={viewStyles}
      onStartShouldSetResponder={() => true}
      onResponderRelease={handleTouch}
      testID="touchableView">
      {Icon}
      <Animated.Text numberOfLines={1} style={[styles.labelText]}>
        {label}
      </Animated.Text>
    </View>
  );
};

const drawerItemDataUnchanged = (
  prevProps: DrawerItemProps,
  nextProps: DrawerItemProps,
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(DrawerItem, drawerItemDataUnchanged);

const styles = StyleSheet.create({
  focusElement: {
    borderColor: COLORS.ORANGE,
    borderWidth: 4,
  },
  selectedElement: {
    borderBottomColor: COLORS.ORANGE,
    borderBottomWidth: 8,
    backgroundColor: COLORS.DARK_GRAY,
  },
  labelText: {
    marginLeft: scaleUxToDp(32.5),
    textAlignVertical: 'center',
    color: COLORS.SMOKE_WHITE,
    fontWeight: '400',
    fontSize: scaleUxToDp(28),
  },
});
