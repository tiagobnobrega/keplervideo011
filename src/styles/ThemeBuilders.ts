import { createThemeFromPartialTheme } from '@amzn/kepler-ui-components';
import { scaleUxToDp } from '../utils/pixelUtils';
import { COLORS } from './Colors';

/**
 * Given the suggested card width, compute all the other dimensions for the
 * card including card height, card child image height/width, margins and
 * so on.
 *
 * @param suggestedCardWidthDP Suggested card width. The width would be clipped
 *    to reside between minimum/maximum width supported by the card.
 * @returns
 */
const computeCardDimensionsForCardWidth = (suggestedCardWidthDP: number) => {
  const maxWidth = scaleUxToDp(440);
  const minWidth = scaleUxToDp(320);
  let cardWidth = suggestedCardWidthDP;

  const cardHeightToWidthRatio = 9 / 16; // 16:9

  let cardHeight = cardWidth * cardHeightToWidthRatio;
  if (cardWidth > maxWidth) {
    cardWidth = maxWidth;
  } else if (cardWidth < minWidth) {
    cardWidth = minWidth;
  }

  const maxHeight = maxWidth * cardHeightToWidthRatio;
  const minHeight = minWidth * cardHeightToWidthRatio;
  if (cardHeight > maxHeight) {
    cardHeight = maxHeight;
  } else if (cardHeight < minHeight) {
    cardHeight = minHeight;
  }

  return {
    card: {
      width: cardWidth,
      height: cardHeight,
    },
  };
};

/**
 * Create the theme with given parameters.
 *
 * @param suggestedCardWidthDP Suggested card width. The width would be clipped
 *    to reside between minimum/maximum width supported by the card.
 */
const createThemeWith = (suggestedCardWidthDP: number) => {
  const cardDimensions =
    computeCardDimensionsForCardWidth(suggestedCardWidthDP);
  console.log('cardDimensions', cardDimensions);

  return createThemeFromPartialTheme({
    card: {
      container: {
        size: {
          width: {
            vertical: {
              md: cardDimensions.card.width,
            },
          },
          height: {
            vertical: {
              md: cardDimensions.card.height,
            },
          },
          borderRadius: {
            sm: 8,
            md: 8,
            lg: 8,
            xl: 8,
          },
        },
        color: {
          borderColor: {
            focused: COLORS.ORANGE,
            pressed: COLORS.ORANGE,
          },
        },
      },
      image: {
        size: {
          margin: 0,
          borderRadius: {
            sm: 0,
            md: 0,
            lg: 0,
            xl: 0,
          },
          width: {
            vertical: {
              md: cardDimensions.card.width,
            },
          },
          height: {
            vertical: {
              md: cardDimensions.card.height,
            },
          },
        },
      },
    },

    typography: {
      text: {
        color: {
          display: COLORS.PALE_BLUE_WHITE,
          headline: COLORS.PALE_BLUE_WHITE,
          title: COLORS.PALE_BLUE_WHITE,
          body: COLORS.PALE_BLUE_WHITE,
          label: COLORS.PALE_BLUE_WHITE,
        },
      },
    },
  });
};

export const createTheme = () => createThemeWith(352);
