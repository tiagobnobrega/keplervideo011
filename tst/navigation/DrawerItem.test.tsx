import { fireEvent, render } from '@testing-library/react-native';
import { isEqual } from 'lodash';
import React from 'react';
import SearchSvg from '../../src/assets/svgr/SearchSVG';
import DrawerItem, {
  DrawerItemProps,
} from '../../src/components/navigation/navigationDrawerContent/DrawerItem';
import { COLORS } from '../../src/styles/Colors';

jest.mock('lodash', () => ({
  isEqual: jest.fn(),
}));

describe('DrawerItem', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (jest.mocked(isEqual) as jest.Mock).mockImplementation((a, b) => {
      return JSON.stringify(a) === JSON.stringify(b);
    });
  });
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  const mockOnPress = jest.fn();

  const defaultProps: DrawerItemProps = {
    isSelected: false,
    isFocused: false,
    isDrawerInFocus: false,
    drawerWidth: 0,
    label: 'Test Label',
    icon: jest.fn(),
    onPress: mockOnPress,
    routeName: 'Profile',
  };
  const renderDrawerItem = (props?: Partial<DrawerItemProps>) => {
    return (
      <DrawerItem
        isSelected={false}
        isFocused={false}
        isDrawerInFocus={false}
        drawerWidth={0}
        label="Test Label"
        routeName="Profile"
        onPress={mockOnPress}
        {...props}
      />
    );
  };
  it('renders correctly', () => {
    const { getByText, toJSON } = render(renderDrawerItem());

    expect(toJSON()).toMatchSnapshot();
    expect(getByText('Test Label')).toBeTruthy();
  });
  it('renders correctly with `isSelected`, `isFocused`, `isDrawerInFocus` enabled', () => {
    const { getByText, toJSON } = render(
      renderDrawerItem({
        isSelected: true,
        isFocused: true,
        isDrawerInFocus: true,
        drawerWidth: 300,
      }),
    );
    expect(toJSON()).toMatchSnapshot();
    expect(getByText('Test Label')).toBeTruthy();
  });
  it('renders correctly with `isSelected`, `isFocused` enabled and `isDrawerInFocus` disabled', () => {
    const { getByText, toJSON } = render(
      renderDrawerItem({
        isSelected: true,
        isFocused: true,
        isDrawerInFocus: false,
        drawerWidth: 0,
      }),
    );
    expect(toJSON()).toMatchSnapshot();
    expect(getByText('Test Label')).toBeTruthy();
  });
  it('renders correctly with `icon` and `isSelected` enabled', () => {
    const { getByText, toJSON } = render(
      renderDrawerItem({
        isSelected: true,
        isFocused: true,
        isDrawerInFocus: false,
        drawerWidth: 0,
        icon: () => {
          return (
            <SearchSvg
              testID="searchsvg"
              width={40}
              height={40}
              stroke={COLORS.SMOKE_WHITE}
              fill={COLORS.SMOKE_WHITE}
            />
          );
        },
      }),
    );
    expect(toJSON()).toMatchSnapshot();
    expect(getByText('Test Label')).toBeTruthy();
  });
  it('renders correctly with `icon` and `isSelected` disabled', () => {
    const { getByText, toJSON } = render(
      renderDrawerItem({
        isSelected: false,
        isFocused: true,
        isDrawerInFocus: false,
        drawerWidth: 0,
        icon: () => {
          return (
            <SearchSvg
              testID="searchsvg"
              width={40}
              height={40}
              stroke={COLORS.SMOKE_WHITE}
              fill={COLORS.SMOKE_WHITE}
            />
          );
        },
      }),
    );
    expect(toJSON()).toMatchSnapshot();
    expect(getByText('Test Label')).toBeTruthy();
  });

  it('does not re-render when props are unchanged', () => {
    const { rerender } = render(renderDrawerItem());

    rerender(renderDrawerItem());

    expect(isEqual).not.toHaveBeenCalledWith(defaultProps, defaultProps);
  });

  it('should call handleTouch and onPress when the view is touched', () => {
    const routeName = 'Profile';
    const { getByTestId } = render(renderDrawerItem());
    const view = getByTestId('touchableView');
    fireEvent(view, 'responderRelease', { preventDefault: jest.fn() });
    expect(mockOnPress).toHaveBeenCalledWith(routeName);
  });
});
