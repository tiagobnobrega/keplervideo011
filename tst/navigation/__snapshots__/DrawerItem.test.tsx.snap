// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerItem renders correctly 1`] = `
<View
  onResponderRelease={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "padding": 20,
        "width": -20,
      },
    ]
  }
  testID="touchableView"
>
  <Text
    collapsable={false}
    numberOfLines={1}
    style={
      Object {
        "color": "#F1F1F1",
        "fontSize": 28,
        "fontWeight": "400",
        "marginLeft": 32.5,
        "textAlignVertical": "center",
      }
    }
  >
    Test Label
  </Text>
</View>
`;

exports[`DrawerItem renders correctly with \`icon\` and \`isSelected\` disabled 1`] = `
<View
  onResponderRelease={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "padding": 20,
        "width": -20,
      },
    ]
  }
  testID="touchableView"
>
  <Text
    collapsable={false}
    numberOfLines={1}
    style={
      Object {
        "color": "#F1F1F1",
        "fontSize": 28,
        "fontWeight": "400",
        "marginLeft": 32.5,
        "textAlignVertical": "center",
      }
    }
  >
    Test Label
  </Text>
</View>
`;

exports[`DrawerItem renders correctly with \`icon\` and \`isSelected\` enabled 1`] = `
<View
  onResponderRelease={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "padding": 20,
        "width": -20,
      },
      Object {
        "backgroundColor": "#1C1C1C",
        "borderBottomColor": "#FFC166",
        "borderBottomWidth": 8,
      },
    ]
  }
  testID="touchableView"
>
  <Text
    collapsable={false}
    numberOfLines={1}
    style={
      Object {
        "color": "#F1F1F1",
        "fontSize": 28,
        "fontWeight": "400",
        "marginLeft": 32.5,
        "textAlignVertical": "center",
      }
    }
  >
    Test Label
  </Text>
</View>
`;

exports[`DrawerItem renders correctly with \`isSelected\`, \`isFocused\` enabled and \`isDrawerInFocus\` disabled 1`] = `
<View
  onResponderRelease={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "padding": 20,
        "width": -20,
      },
      Object {
        "backgroundColor": "#1C1C1C",
        "borderBottomColor": "#FFC166",
        "borderBottomWidth": 8,
      },
    ]
  }
  testID="touchableView"
>
  <Text
    collapsable={false}
    numberOfLines={1}
    style={
      Object {
        "color": "#F1F1F1",
        "fontSize": 28,
        "fontWeight": "400",
        "marginLeft": 32.5,
        "textAlignVertical": "center",
      }
    }
  >
    Test Label
  </Text>
</View>
`;

exports[`DrawerItem renders correctly with \`isSelected\`, \`isFocused\`, \`isDrawerInFocus\` enabled 1`] = `
<View
  onResponderRelease={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 10,
        "marginRight": 10,
        "padding": 20,
        "width": 280,
      },
      Object {
        "borderColor": "#FFC166",
        "borderWidth": 4,
      },
    ]
  }
  testID="touchableView"
>
  <Text
    collapsable={false}
    numberOfLines={1}
    style={
      Object {
        "color": "#F1F1F1",
        "fontSize": 28,
        "fontWeight": "400",
        "marginLeft": 32.5,
        "textAlignVertical": "center",
      }
    }
  >
    Test Label
  </Text>
</View>
`;
